### PR TITLE
fix: electron update error

### LINKED ISSUE #5551: Data race in auto updating vs manual updating
## Describe the bug

We have a data race when user click "check for updates" with auto-update set to true, causing the updating fail (#5549). 

The reason is when auto-update is set to true
1. when composer starts up, we are checking the updates at backend and starts the downloading process, a temp file is created for the
2. when user click check for updates, we are trying to write a same temp file at the same location, then this operation fails because the permission is not allowed, since this file is already created and been actively used. 
3. when the operation in 2 fails, the clean process is trigger, and this file is totally removed, causing composer to be in a bad state where any other following auto & manual update won't work

## Version

lastest main
## Browser

- [x] Electron distribution
- [ ] Chrome
- [ ] Safari
- [ ] Firefox
- [ ] Edge

## OS

- [ ] macOS
- [x] Windows
- [ ] Ubuntu

## To Reproduce

1. Install 1.3.0, turn on auto update
2. Run composer, click "help" => "check for updates"
3. See error

## Expected behavior

1. When auto update is on and is downloading, and when user click manual update, we should prompting user to show we are already doing a downloading now.
2. When auto update finished it's downloading, and user should get a notification to ask user to update or not
3. If user choose to ignore the notification. Then when user click manual update, we will show the same notification again asking for update, without have to download again (if there is not new version).

### PR DESCRIPTION
**Pull request by zhixzhan**
## Description
We have a data race when user click "check for updates" with auto-update set to true, causing the updating fail (#5549).

This PR let manual-update can continue auto-update progress.

1. if auto update is downloading, when click "check for updates" will show the progress.
2. if auto update is downloaded, when click "check for updates" will show quit-and-start modal.

## Task Item
close #5551 

## Screenshots

1. When auto update is downloading.

2. When auto update is downloaded.

### CODE REVIEW

File: Composer/packages/electron-server/src/appUpdater.ts [hunk 1]
```diff
@@ -108,6 +111,7 @@ export class AppUpdater extends EventEmitter {
 
   private onDownloadProgress(progress: any) {
     log('Got update progress: %O', progress);
+    this.downloadingUpdate = true;
```
Comment (zhixzhan) [inline Composer/packages/electron-server/src/appUpdater.ts:114]: Auto update won't call `public downloadUpdate()`,  So seems here is the only place we can know current is downloading.

File: Composer/packages/electron-server/src/appUpdater.ts [hunk 2]
```diff
@@ -116,10 +120,10 @@ export class AppUpdater extends EventEmitter {
   private onUpdateDownloaded(updateInfo: UpdateInfo) {
     log('Update downloaded: %O', updateInfo);
     this._downloadedUpdate = true;
-    this.resetToIdle();
     if (this.explicitCheck || !this.settings.autoDownload) {
       this.emit('update-downloaded', updateInfo);
     }
+    this.resetToIdle();
```
Comment (zhixzhan) [inline Composer/packages/electron-server/src/appUpdater.ts:126]: reset will clean up `this.explictCheck`, so here place it in the end.

File: Composer/packages/client/src/components/AppUpdater.tsx [hunk 3]
```diff
@@ -131,6 +131,12 @@ export const AppUpdater: React.FC<{}> = () => {
           break;
         }
 
+        case 'update-in-progress': {
+          setAppUpdateStatus(AppUpdaterStatus.UPDATE_AVAILABLE, undefined);
```
Comment (tonyanziano) [inline Composer/packages/client/src/components/AppUpdater.tsx:135]: the `undefined` argument here is supposed to be the version of the update -- if you look at the first video you posted, you can see that when this line is called and shows the "Update Available" dialog, the dialog shows "vundefined"

image

**Review by tonyanziano — CHANGES_REQUESTED** [review]
Aside from the visual bug in my comment, I think this looks good! Nice job

File: Composer/packages/client/src/components/AppUpdater.tsx (hunk 3 above)
Comment (zhixzhan) [reply Composer/packages/client/src/components/AppUpdater.tsx]: Nice catch!
I see the differences, now updated: 

image

**Review by tonyanziano — APPROVED** [review]

**Pull request merged**